pkg/chains/factory.go

```diff
@@ -0,0 +1,144 @@
+			if err != nil {
+				l.Warnf("error configuring x509 signer: %s", err)
+				continue
+			}
+			all[s] = signer
+		case signing.TypeKMS:
+			signer, err := kms.NewSigner(ctx, cfg.Signers.KMS, l)
+			if err != nil {
+				l.Warnf("error configuring kms signer with config %v: %s", cfg.Signers.KMS, err)
+				continue
+			}
+			all[s] = signer
+		default:
+			// This should never happen, so panic
+			l.Panicf("unsupported signer: %s", s)
+		}
+	}
+	return all
+}
+
+func allFormatters(cfg config.Config, l *zap.SugaredLogger) map[formats.PayloadType]formats.Payloader {
+	all := map[formats.PayloadType]formats.Payloader{}
+
+	for _, f := range formats.AllFormatters {
+		switch f {
+		case formats.PayloadTypeTekton:
+			formatter, err := tekton_format.NewFormatter()
+			if err != nil {
+				l.Warnf("error configuring tekton formatter: %s", err)
+			}
+			all[f] = formatter
+		case formats.PayloadTypeSimpleSigning:
+			formatter, err := simple.NewFormatter()
+			if err != nil {
+				l.Warnf("error configuring simplesigning formatter: %s", err)
+			}
+			all[f] = formatter
+		case formats.PayloadTypeInTotoIte6:
+			formatter, err := intotoite6.NewFormatter(cfg, l)
+			if err != nil {
+				l.Warnf("error configuring intoto formatter: %s", err)
+			}
+			all[f] = formatter
+		case formats.PayloadTypeProvenance:
+			formatter, err := provenance.NewFormatter(cfg, l)
+			if err != nil {
+				l.Warnf("error configuring tekton-provenance formatter: %s", err)
+			}
+			all[f] = formatter
+		}
+	}
+
+	return all
+}
+
+// allBackends creates and initializes every configured storage backend.
+func allBackends(ctx context.Context, ps versioned.Interface, kc kubernetes.Interface, logger *zap.SugaredLogger, cfg config.Config) (map[string]storage.Backend, error) {
+	// Add an entry here for every configured backend
+	configuredBackends := []string{}
+	if cfg.Artifacts.TaskRuns.Enabled() {
+		configuredBackends = append(configuredBackends, cfg.Artifacts.TaskRuns.StorageBackend.List()...)
+	}
+	if cfg.Artifacts.OCI.Enabled() {
+		configuredBackends = append(configuredBackends, cfg.Artifacts.OCI.StorageBackend.List()...)
+	}
+
+	// Now only initialize and return the configured ones.
+	backends := map[string]storage.Backend{}
+	for _, backendType := range configuredBackends {
+		switch backendType {
+		case gcs.StorageBackendGCS:
+			gcsBackend, err := gcs.NewStorageBackend(ctx, logger, cfg)
+			if err != nil {
+				return nil, err
+			}
+			backends[backendType] = gcsBackend
+		case tekton_storage.StorageBackendTekton:
+			backends[backendType] = tekton_storage.NewStorageBackend(ps, logger)
+		case oci.StorageBackendOCI:
+			ociBackend := oci.NewStorageBackend(ctx, logger, kc, cfg)
+			backends[backendType] = ociBackend
+		case docdb.StorageTypeDocDB:
+			docdbBackend, err := docdb.NewStorageBackend(ctx, logger, cfg)
+			if err != nil {
+				return nil, err
+			}
+			backends[backendType] = docdbBackend
+		}
+	}
+	return backends, nil
+}
```

Comment:
Yeah that's a totally reasonable concern though I didn't change the implementation of `allSigners`, `allFormatters ` (from signing.go), or `allBackends` (from storage.go though the name is changed from `InitializeBackends` to `allBackends` for consistency).
The reason why I pull them out is that I though they will be created per process level so it might make more sense to put them in factory.go. But it makes sense to me to have them in  `signing.go` too.

Do we want to put all three functions back to where they were, and call them separately in controller.go instead of using a single factory function?